### PR TITLE
Adding t5 inference to the example 

### PR DESCRIPTION
**Pull request by HamidShojanazeri**
Starting to explore Pippy as distributed inference solution for PyTorch. This PR will work through this T5 example to get it working with the auto-split(model, segments) API and benchmark it for inference.

### CODE REVIEW

File: examples/hf/t5/pippy_t5.py [hunk 1]
```diff
@@ -115,6 +123,70 @@ def resolve_pg_per_stage(pp_rank):
 
 
 def run_master(pp_ranks, args):
+
+    def train_mode(model,split_policy, MULTI_USE_PARAM_CONFIG , concrete_args):
```
Comment (anj-s) [inline examples/hf/t5/pippy_t5.py:127]: Can these be functions outside run_master? They seem independent enough.

Comment (anj-s) [inline examples/hf/t5/pippy_t5.py:127]: nit: space after "," and no space after params.

File: examples/hf/t5/pippy_t5.py [hunk 2]
```diff
@@ -115,6 +123,70 @@ def resolve_pg_per_stage(pp_rank):
...
+
+        output_loss_value_spec = {'loss': True, 'logits': False,
+                                # 'past_key_values': False,
```
Comment (anj-s) [inline examples/hf/t5/pippy_t5.py:130]: what is "past_key_values" for?

File: examples/hf/t5/pippy_t5.py [hunk 3]
```diff
@@ -154,52 +227,27 @@ def run_master(pp_ranks, args):
...
-    t5_input_dict = {'input_ids': inp, 'decoder_input_ids': inp,
-                     'labels': torch.empty(bs, seq_length, dtype=torch.long, device=device).random_(
-                         t5.config.vocab_size - 1)}
+
+    if args.mode == "train":
+        t5_input_dict = {'input_ids': inp, 'decoder_input_ids': inp,
+                        'labels': torch.empty(bs, seq_length, dtype=torch.long, device=device).random_(
+                            t5.config.vocab_size - 1)}
+    elif args.mode == "inference":
+        t5_input_dict = {'input_ids': inp, 'decoder_input_ids': inp}
```
Comment (anj-s) [inline examples/hf/t5/pippy_t5.py:236]: where do we load model checkpoints for inference?

File: examples/hf/t5/pippy_t5.py [hunk 4]
```diff
@@ -37,6 +37,14 @@
 
 pippy.fx.Tracer.proxy_buffer_attributes = True
 
+def model_size(model_pipe):
+
+    total_params = 0
+    for i, sm in enumerate(model_pipe.split_gm.children()):
+        params = get_number_of_params(sm)
+        print(f"submod_{i} {params // 10 ** 6}M params")
```
[the remark this answers is not in the document]
Comment (HamidShojanazeri) [reply examples/hf/t5/pippy_t5.py:45]: I think as this is a simple example, prints could help to keep it educational.

File: examples/hf/t5/pippy_t5.py (hunk 3 above)
Comment (HamidShojanazeri) [reply examples/hf/t5/pippy_t5.py:236]: They are both using the same checkpoints, in this case its not a real checkpoint, it is mostly to demonstrate functionality other than that there is no real data going toward training/ inference.

File: examples/hf/t5/pippy_t5.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (HamidShojanazeri) [reply examples/hf/t5/pippy_t5.py:56]: Regarding the prints, It was from there from the original implementation, I think as the purpose of the example is just educational that might help to keep track of what is happening, but for sure we can change them to be part of the loggings.

**Comment by kwen2501** [comment]
- Further merging functions `train_mode` and `inference_mode` into one function (named `transform_into_pipeline`)
- `args.mode` changed to `args.train = 0|1`

**Comment by kwen2501** [comment]
Inference:
`python pippy_t5.py --train 0`

Training:
`python pippy_t5.py --train 1`

**Review by kwen2501 — APPROVED** [review]

**Pull request merged**